### ISSUE TITLE
Train Test and Validation Sets

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8829]
Train, Test, and Validation Sets<br>
<br>
https://ift.tt/UeFSghu<br>
<br>
Jared Wilber<br>
<br>
<div>
<div>
<div>
<div>
<article>
<section>
<h2>The Importance of Data Splitting<br>
</h2>
<p>By <a href="https://twitter.com/jdwlbr">Jared Wilber</a> &amp; Brent Werness<br>
<br>
<br>
<br>
In most supervised machine learning tasks, best practice recommends to split your data into three independent sets: a <span><b>training set</b></span>, a <span><b>testing set</b></span>, and a <span><b>validation set</b></span>.<br>
<br>
To learn why, let's pretend that we have a dataset of two types of pets:<br>
<br>
<span><span>Cats</span>:&nbsp; <img src="https://mlu-explain.github.io/train-test-validation/noun_Cat_18061.575745a8.svg">&nbsp;&nbsp;&nbsp;&nbsp; <span>Dogs</span>:&nbsp; <img src="https://mlu-explain.github.io/train-test-validation/noun_Dog_79225.bf4c88d0.svg"></span><br>
<br>
Each pet in our dataset has two features: <span>weight</span> and <span>fluffiness</span>.<br>
<br>
Our goal is to identify and evaluate suitable models for classifying a given pet as either a cat or a dog. We'll use train/test/validations splits to do this!</p>
</section>
<section>
<h2>Train, Test, and Validation Splits</h2>
<p>The first step in our classification task is to randomly split our pets into three independent sets:<br>
<br>
<span><span>Training Set</span></span>: <span>The dataset that we feed our model to learn potential underlying patterns and relationships.</span><br>
<br>
<span><span>Validation Set</span></span>: <span>The dataset that we use to understand our model's performance across different model types and hyperparameter choices.</span><br>
<br>
<span><span>Test Set</span></span>: <span>The dataset that we use to approximate our model's unbiased accuracy in the wild.</span></p>
</section>
<section>
<h2>The Training Set</h2>
<p><span>The training set is the dataset that we employ to train our model.</span> It is this dataset that our model uses to learn any underlying patterns or relationships that will enable making predictions later on.<br>
<br>
The training set should be as representative as possible of the population that we are trying to model. Additionally, we need to be careful and ensure that it is as unbiased as possible, as any bias at this stage may be propagated downstream during inference.</p>
</section>
<section>
<h2>Building Our Model</h2>
<p>Our goal (to determine whether a given pet is a cat or a dog) is a binary classification task, so we will use a simple but effective model appropriate for this task: <a href="https://www.youtube.com/watch?v=jhJwNpidiqM"><span>logistic regression</span></a>.<br>
<br>
Logistic regression will learn a decision boundary to best separate the cats from dogs in our training data, using the selected feature (<span>None</span>, <span>Weight</span>, <span>Fluffiness</span>, or <i>both</i> <span>Weight</span> and <span>Fluffiness</span>).<br>
<br>
Select the feature to visualize the corresponding logistic regression model's decision boundary. <b>Drag each animal in the training set to a new position to see how the boundary updates!</b></p>
</section>
<section>
<h2>The Validation Set</h2>
<p>We can build four different logistic regression models (one for each feature possibility), <b>how do we decide which model to select?</b><br>
<br>
We could compare the accuracy of each model on the training set, but if we use the same exact dataset for both training and tuning, the model will overfit and won't generalize well.<br>
<br>
<b>This is where the validation set comes in</b> &mdash; <span>it acts as an independent, unbiased dataset for comparing the performance of different algorithms trained on our training set.</span><br>
<br>
Select a feature to view the model's performance on the validation set in the table below. <b>Drag the pets across the line to see how the model performance updates!</b></p>
</section>
<section>
<h2>The Testing Set</h2>
<p>Once we have used the validation set to determine the algorithm and parameter choices that we would like to use in production, the test set is used to approximate the models's true performance in the wild. <span>It is the final step in evaluating our model's performance on unseen data.</span><br>
<br>
<b>We should never, under any circumstance, look at the test set's performance before selecting a model.</b><br>
<br>
Peeking at our test set performance ahead of time is a form of overfitting, and will likely lead to unreliable performance expectations in production. It should only be checked as the final form of evaluation, after the validation set has been used to identify the best model.</p>
</section>
<section>
<h2>Summary</h2>
<p>You may have noticed that the test accuracy of the "just fluffiness" model was higher than that of the "both features" model, despite the validation set selecting the latter model as the best. This occurrence of the validation performance not exactly matching the test performance might happen, yet it is not a bad thing. Remember that the test performance is not a number to optimize over &mdash; it is a metric to assess future performance. It allows us to estimate, with confidence, that our model can distinguish between cats and dogs with 87.5% accuracy.<br>
<br>
<span>Key Takeaways</span><br>
<br>
It is best practice in machine learning to split our data into the following three groups:<br>
<br>
<span><span>Training Set</span>: <span>For training of the model.</span></span><br>
<br>
<span><span>Validation Set</span>: <span>For unbiased evaluation of the model.</span></span><br>
<br>
<span><span>Test Set</span>: <span>For final evaluation of the model.</span></span><br>
<br>
Adhering to this setup helps ensure that we have a realistic understanding of our model's performance, and that we (hopefully) built a model that generalizes well to unseen data.<br>
<br>
<span>&#128062;</span><br>
<br>
Thanks for reading! To learn more about machine learning, check out <a href="https://aws.amazon.com/machine-learning/mlu/">our website</a>, watch <a href="https://www.youtube.com/channel/UC12LqyqTQYbXatYS9AA7Nuw">our videos</a>, or read the <a href="https://d2l.ai/">D2L book</a>.<br>
<br>
Animal icons by Adrien Coquet (cat) &amp; Maur&iacute;cio Brito (dog).<br>
<br></p>
</section>
</article>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via MLU-Explain<br>
<br>
September 21, 2022 at 12:06AM
